linux/embedded_rootfs/pkg_addon/strongswan-4.0.4/src/libfreeswan/ttodata.3
.TH IPSEC_TTODATA 3 "16 August 2003"
.\" RCSID $Id: ttodata.3,v 1.1 2007-06-25 11:14:07 encounter Exp $
.SH NAME
ipsec ttodata, datatot \- convert binary data bytes from and to text formats
.SH SYNOPSIS
.B "#include <freeswan.h>"
.sp
.B "const char *ttodata(const char *src, size_t srclen,"
.ti +1c
.B "int base, char *dst, size_t dstlen, size_t *lenp);"
.br
.B "const char *ttodatav(const char *src, size_t srclen,"
.ti +1c
.B "int base, char *dst, size_t dstlen, size_t *lenp,"
.ti +1c
.B "char *errp, size_t errlen, int flags);"
.br
.B "size_t datatot(const char *src, size_t srclen,"
.ti +1c
.B "int format, char *dst, size_t dstlen);"
.SH DESCRIPTION
.IR Ttodata ,
.IR ttodatav ,
and
.I datatot
convert arbitrary binary data (e.g. encryption or authentication keys)
from and to more-or-less human-readable text formats.
.PP
Currently supported formats are hexadecimal, base64, and characters.
.PP
A hexadecimal text value begins with a
.B 0x
(or
.BR 0X )
prefix and continues with two-digit groups
of hexadecimal digits (0-9, and a-f or A-F),
each group encoding the value of one binary byte, high-order digit first.
A single
.B _
(underscore)
between consecutive groups is ignored, permitting punctuation to improve 
readability; doing this every eight digits seems about right.
.PP
A base64 text value begins with a
.B 0s
(or
.BR 0S )
prefix 
and continues with four-digit groups of base64 digits (A-Z, a-z, 0-9, +, and /),
each group encoding the value of three binary bytes as described in
section 6.8 of RFC 2045.
If
.B flags
has the
.B TTODATAV_IGNORESPACE
bit on, blanks are ignore (after the prefix).
Note that the last one or two digits of a base64 group can be
.B =
to indicate that fewer than three binary bytes are encoded.
.PP
A character text value begins with a
.B 0t
(or
.BR 0T )
prefix
and continues with text characters, each being the value of one binary byte. 
.PP
All these functions basically copy data from
.I src
(whose size is specified by
.IR srclen )
to
.I dst
(whose size is specified by
.IR dstlen ),
doing the conversion en route.
If the result will not fit in
.IR dst ,
it is truncated;
under no circumstances are more than
.I dstlen
bytes of result written to
.IR dst .
.I Dstlen
can be zero, in which case
.I dst
need not be valid and no result bytes are written at all.
.PP
The
.I base
parameter of
.I ttodata
and
.I ttodatav
specifies what format the input is in;
normally it should be
.B 0
to signify that this gets figured out from the prefix.
Values of
.BR 16 ,
.BR 64 ,
and
.BR 256
respectively signify hexadecimal, base64, and character-text formats
without prefixes.
.PP
The
.I format
parameter of
.IR datatot ,
a single character used as a type code,
specifies which text format is wanted.
The value
.B 0
(not ASCII
.BR '0' ,
but a zero value) specifies a reasonable default.
Other currently-supported values are:
.RS 2
.TP 4
.B 'x'
continuous lower-case hexadecimal with a
.B 0x
prefix
.TP
.B 'h'
lower-case hexadecimal with a
.B 0x
prefix and a
.B _
every eight digits
.TP
.B ':'
lower-case hexadecimal with no prefix and a
.B :
(colon) every two digits
.TP
.B 16
lower-case hexadecimal with no prefix or
.B _
.TP
.B 's'
continuous base64 with a
.B 0s
prefix
.TP
.B 64
continuous base64 with no prefix
.RE
.PP
The default format is currently
.BR 'h' .
.PP
.I Ttodata
returns NULL for success and
a pointer to a string-literal error message for failure;
see DIAGNOSTICS.
On success,
if and only if
.I lenp
is non-NULL,
.B *lenp
is set to the number of bytes required to contain the full untruncated result.
It is the caller's responsibility to check this against
.I dstlen
to determine whether he has obtained a complete result.
The
.B *lenp
value is correct even if
.I dstlen
is zero, which offers a way to determine how much space would be needed
before having to allocate any.
.PP
.I Ttodatav
is just like
.I ttodata
except that in certain cases,
if
.I errp
is non-NULL,
the buffer pointed to by
.I errp
(whose length is given by
.IR errlen )
is used to hold a more detailed error message.
The return value is NULL for success,
and is either
.I errp
or a pointer to a string literal for failure.
If the size of the error-message buffer is
inadequate for the desired message,
.I ttodatav
will fall back on returning a pointer to a literal string instead.
The
.I freeswan.h
header file defines a constant
.B TTODATAV_BUF
which is the size of a buffer large enough for worst-case results.
.PP
The normal return value of
.IR datatot
is the number of bytes required
to contain the full untruncated result.
It is the caller's responsibility to check this against
.I dstlen
to determine whether he has obtained a complete result.
The return value is correct even if
.I dstlen
is zero, which offers a way to determine how much space would be needed
before having to allocate any.
A return value of
.B 0
signals a fatal error of some kind
(see DIAGNOSTICS).
.PP
A zero value for
.I srclen
in
.I ttodata
(but not
.IR datatot !)
is synonymous with
.BR strlen(src) .
A non-zero
.I srclen
in
.I ttodata
must not include the terminating NUL.
.PP
Unless
.I dstlen
is zero,
the result supplied by
.I datatot
is always NUL-terminated,
and its needed-size return value includes space for the terminating NUL.
.PP
Several obsolete variants of these functions
.RI ( atodata ,
.IR datatoa ,
.IR atobytes ,
and
.IR bytestoa )
are temporarily also supported.
.SH SEE ALSO
sprintf(3), ipsec_atoaddr(3)
.SH DIAGNOSTICS
Fatal errors in
.I ttodata
and
.I ttodatav
are:
unknown characters in the input;
unknown or missing prefix;
unknown base;
incomplete digit group;
non-zero padding in a base64 less-than-three-bytes digit group;
zero-length input.
.PP
Fatal errors in
.I datatot
are:
unknown format code;
zero-length input.
.SH HISTORY
Written for the FreeS/WAN project by Henry Spencer.
.SH BUGS
.I Datatot
should have a format code to produce character-text output.
.PP
The
.B 0s
and
.B 0t
prefixes are the author's inventions and are not a standard
of any kind.
They have been chosen to avoid collisions with existing practice
(some C implementations use
.B 0b
for binary)
and possible confusion with unprefixed hexadecimal.


linux/release-notes.txt
Release notes for Octeon SDK CN3XXX-LINUX-1.5.0

Release 1.5.0
	- Embedded root filesystem:
		- Add module-init-tools
		- Updated testsuite
		- Do a modules_prepare to get build setup for new config files
		- Add Oprofile
		- Fix path for iptables finding its shared libraries
		- Add cpu-load script for getting load percent
		- Uclibc support
		- NFS root: Fix path to ethernet driver and chroot
		- Load IPv6 kernel module if it exists
		- Fix bug where fwd rule not supported by ipsec-tools
		- Improve logic on stripping to only strip ELF files
		- Add fold for wrapping lines
		- Bug 799: Disable crypto in tcpdump with --without-crypto
		- Calculate the size of the ext3 filesystem instead of
			hardcoding a value
		- Move modules into /lib/modules/<kernel version>
		- Add iproute2 package for IPv6 support
		- Add oct-linux-profile
		- Add missing cramfs in help message
		- Add global option to turn on/off IPv6
		- Add ethtool
		- Add distclean target and make oldconfig silent
		- Fix bug where mii-tool isn't installed
		- Include user-include if it exists and a different dir
			wasn't specified. This makes it easy to add files
			without affecting the rootfs config. It is also
			more like the old minimal_rootfs.
		- Update OpenSSL version from 0.9.8b to 0.9.8d
		- Fix OpenSSL function typecast incompatibility with GCC 4.x
	- Ethernet driver:
		- Add NETPOLL support so we can support netconsole
		- Add define DONT_WRITEBACK to control if FPA frees also tell
			the L2 to throw away changes to the buffer
		- Set carrier status when link goes up or down
		- Provide help information for module parameters
		- Only enable the ethtool interface on RGMII ports
		- Fix long standing bug where RGMII ports are assigned the
			wrong PHY_ID if they're on interface 2 after an
			enabled SPI.
		- Increment RX counters for virtual ports
		- Don't use bootloader MAC addresses for virtual ports
		- Add workaround if bootloader mistakenly leaves GMX enabled
		- Set the MTU before the device comes up. It was defaulting to
			10k
		- Added parameter to skip SPI interface initialization
		- Large rewrite of the RGMII speed code. This code works for
			plug and unplug. It has trouble with unplug/plug with
			half duplex.
		- Fix crash when bridging. Don't put shared/cloned packets in
			the HW pool.
		- Remove reference to non-existent skb->tc_classid
		- Enable backpressure for SPI and RGMII
		- Set NETIF_F_LLTX - We do our own locking, Linux doesn't need
			to. This gives us a major performance boost
		- Add a new option (CVMX_ENABLE_LEN_M8_FIX) for enabling
			hardware fix for errata PKI-100
		- Cleanup simulator support. Don't do RGMII autonegotiation
			and don't read/write unmodeled CSRs.
		- Fix Bug 804. Non aligned buffers were being put in the FPA,
			causing memory corruption.
	- Intercept Module Example
		- Don't crash if ethernet driver isn't loaded. Just print a
			message and stop loading.
	- Add DRAM utilization to /proc/octeon_perf
	- Add sysn32_rt_sigtimedwait back that was removed in 2.6.16. Fixes
		N32 sigtimedwait
	- Change blocksize to 4 in NULL crypto cypher so performance isn't
		quite so bad for kernel IPSEC null encryption.
	- Add CN58XX simulator support.
	- Add CN56XX simulator support.
	- TWSI driver: Fixed problem with multi-byte read transfers (bug#793)
	- Kernel memcpy: Fix bug where prefetches were disabled.
	- Implement set_affinity for IRQs
	- PCI: Updated PCI bus bridge initialization
	- PCI: Fix PCI bug with Intel bridges causing poor performance. Intel bridges
		interpret a cache line size of 128 as an error case. Changed
		PCI cache line size to 64 to avoid this problem.			
	- PCI: Enable error detection and propagation on all PCI bridges.
	- PCI: Configure arbiter to use round robin priority instead of fixed.
		This should prevent a single device from hogging the bus.
	- PCI: Improve EBH3000 interrupt routing
	- PCI: Fix DMA mask determination in dma_map_single to correctly use
		dev->dma_mask. Fix SATA Sil 3124 driver to set DMA mask
		properly. As a side effect, the E1000 is now picked up as a
		64bit device. This fixes the problem with BAR mapping during
		unload.
	- PCI: Eliminate reads from Bar1 index registers on dma_map/unmap.
		This should be much faster.
	- PCI: Enabling smart prefetch and divided fifos for PLX 6540 bridge
	- PCI: Move BAR2 mapping code before BAR0/1 so there isn't ever an
		overlap. Clear pending PCI error interrupts after init.
	- PCI: Add workaround for PCI-302 errata
	- Debian: Refuse to destroy disks larger than 4GB
	- Debian glibc: Update to add missing .set mips0 after code Octeon added
	- Debian: Add uClibc libaries to file system.
	- Implementation of fast TLS access. See the SDK documentation for more
		information.
	- Fix Bug where a RLM interrupt on CN30XX would read Zip block, causing
		strange failures
	- Use sparsemem static instead of extreme. The extreme version uses some
		very early mem alloc that have addressing requirements. These
		can randomly corrupt low memory with any sign. The static
		version is just as good for us.
	- Add POW only ethernet driver for use with networking between cores
		running different Linux kernels. Driver is GPL.
	- CIU interrupt processing assembly was optimized to remove loops.
	- Add config options to lock often used kernel code in the L2
	- Added Octeon specific cache error exception handler.
	- Remove writing zero to L2C perf counters. They are read only.
	- Memcpy: Minor optimizations to save a few cycles
	- Use runtime checks in modules. Undefine OCTEON_MODEL so we'll get
		warnings and errors when it is used.
	- Rewrite RML interrupt handling code to handle all interrupts except
		for GXM and SPI. These should be done in the ethernet driver.
	- Add the natsemi PCI driver to the kernel config
	- Enable IPsec as modules
	- Fix oversight: enable the perfcounter interrupt. This caused Oprofile
		results to be incorrect.

linux/embedded_rootfs/pkg_addon/strongswan-4.0.4/src/libstrongswan/asn1/ttodata.c
/*
 * convert from text form of arbitrary data (e.g., keys) to binary
 * Copyright (C) 2000  Henry Spencer.
 * 
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/lgpl.txt>.
 * 
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
 * License for more details.
 */

#include "ttodata.h"

#include <string.h>
#include <ctype.h>

/* converters and misc */
static int unhex(const char *, char *, size_t);
static int unb64(const char *, char *, size_t);
static int untext(const char *, char *, size_t);
static const char *badch(const char *, int, char *, size_t);

/* internal error codes for converters */
#define	SHORT	(-2)		/* internal buffer too short */
#define	BADPAD	(-3)		/* bad base64 padding */
#define	BADCH0	(-4)		/* invalid character 0 */
#define	BADCH1	(-5)		/* invalid character 1 */
#define	BADCH2	(-6)		/* invalid character 2 */
#define	BADCH3	(-7)		/* invalid character 3 */
#define	BADOFF(code) (BADCH0-(code))

/**
 * @brief convert text to data, with verbose error reports
 * 
 * If some of this looks slightly odd, it's because it has changed
 * repeatedly (from the original atodata()) without a major rewrite.
 *
 * @param src
 * @param srclen	0 means apply strlen()
 * @param base 		0 means figure it out
 * @param dst 		need not be valid if dstlen is 0
 * @param dstlen	
 * @param lenp		where to record length (NULL is nowhere)
 * @param errp		error buffer
 * @param flags
 * @return			NULL on success, else literal or errp
 */
const char *ttodatav(const char *src, size_t srclen, int base, char *dst, size_t dstlen, size_t *lenp, char *errp, size_t errlen, unsigned int flags)
{
	size_t ingroup;	/* number of input bytes converted at once */
	char buf[4];		/* output from conversion */
	int nbytes;		/* size of output */
	int (*decode)(const char *, char *, size_t);
	char *stop;
	int ndone;
	int i;
	int underscoreok;
	int skipSpace = 0;

	if (srclen == 0)
		srclen = strlen(src);
	if (dstlen == 0)
		dst = buf;	/* point it somewhere valid */
	stop = dst + dstlen;

	if (base == 0) {
		if (srclen < 2)
			return "input too short to be valid";
		if (*src++ != '0')
			return "input does not begin with format prefix";
		switch (*src++) {
		case 'x':
		case 'X':
			base = 16;
			break;
		case 's':
		case 'S':
			base = 64;
			break;
		case 't':
		case 'T':
			base = 256;
			break;
		default:
			return "unknown format prefix";
		}
		srclen -= 2;
	}
	switch (base) {
	case 16:
		decode = unhex;
		underscoreok = 1;
		ingroup = 2;
		break;
	case 64:
		decode = unb64;
		underscoreok = 0;
		ingroup = 4;
		if(flags & TTODATAV_IGNORESPACE) {
			skipSpace = 1;
		}
		break;

	case 256:
		decode = untext;
		ingroup = 1;
		underscoreok = 0;
		break;
	default:
		return "unknown base";
	}

	/* proceed */
	ndone = 0;
	while (srclen > 0) {
		char stage[4];	/* staging area for group */
		size_t sl = 0;

		/* Grab ingroup characters into stage,
		 * squeezing out blanks if we are supposed to ignore them.
		 */
		for (sl = 0; sl < ingroup; src++, srclen--) {
			if (srclen == 0)
				return "input ends in mid-byte, perhaps truncated";
			else if (!(skipSpace && (*src == ' ' || *src == '\t')))
				stage[sl++] = *src;
		}
		
		nbytes = (*decode)(stage, buf, sizeof(buf));
		switch (nbytes) {
		case BADCH0:
		case BADCH1:
		case BADCH2:
		case BADCH3:
			return badch(stage, nbytes, errp, errlen);
		case SHORT:
			return "internal buffer too short (can't happen)";
		case BADPAD:
			return "bad (non-zero) padding at end of base64 input";
		}
		if (nbytes <= 0)
			return "unknown internal error";
		for (i = 0; i < nbytes; i++) {
			if (dst < stop)
				*dst++ = buf[i];
			ndone++;
		}
		while (srclen >= 1 && skipSpace && (*src == ' ' || *src == '\t')){
			src++;
			srclen--;
		}
		if (underscoreok && srclen > 1 && *src == '_') {
			/* srclen > 1 means not last character */
			src++;
			srclen--;
		}
	}

	if (ndone == 0)
		return "no data bytes specified by input";
	if (lenp != NULL)
		*lenp = ndone;
	return NULL;
}

/**
 * @brief ttodata - convert text to data
 * 
 * @param src
 * @param srclen	0 means apply strlen()
 * @param base		0 means figure it out
 * @param dst		need not be valid if dstlen is 0
 * @param dstlen
 * @param lenp		where to record length (NULL is nowhere)
 * @return			NULL on success, else literal
 */
const char *ttodata(const char *src, size_t srclen, int base, char *dst, size_t dstlen, size_t *lenp)
{
	return ttodatav(src, srclen, base, dst, dstlen, lenp, (char *)NULL,
			(size_t)0, TTODATAV_SPACECOUNTS);
}

/**
 * @brief atodata - convert ASCII to data
 * 
 * backward-compatibility interface
 * 
 * @param src
 * @param srclen
 * @param dst
 * @param dstlen
 * @return 			0 for failure, true length for success
 */
size_t atodata(const char *src, size_t srclen, char *dst, size_t dstlen)
{
	size_t len;
	const char *err;

	err = ttodata(src, srclen, 0, dst, dstlen, &len);
	if (err != NULL)
		return 0;
	return len;
}

/**
 * @brief  atobytes - convert ASCII to data bytes
 *
 * another backward-compatibility interface
 */
const char *atobytes(const char *src, size_t srclen, char *dst, size_t dstlen, size_t *lenp)
{
	return ttodata(src, srclen, 0, dst, dstlen, lenp);
}

/**
 * @brief unhex - convert two ASCII hex digits to byte
 * 
 * @param src 		known to be full length
 * @param dstnumber of result bytes, or error code
 * @param dstlen	not large enough is a failure
 * @return			
 */
static int unhex(const char *src, char *dst, size_t dstlen)
{
	char *p;
	unsigned byte;
	static char hex[] = "0123456789abcdef";

	if (dstlen < 1)
		return SHORT;
	
	p = strchr(hex, *src);
	if (p == NULL)
		p = strchr(hex, tolower(*src));
	if (p == NULL)
		return BADCH0;
	byte = (p - hex) << 4;
	src++;

	p = strchr(hex, *src);
	if (p == NULL)
		p = strchr(hex, tolower(*src));
	if (p == NULL)
		return BADCH1;
	byte |= (p - hex);

	*dst = byte;
	return 1;
}

/**
 * @brief unb64 - convert four ASCII base64 digits to three bytes
 *
 * Note that a base64 digit group is padded out with '=' if it represents
 * less than three bytes:  one byte is dd==, two is ddd=, three is dddd.
 *
 * @param src		known to be full length 
 * @param dst		
 * @param dstlen	
 * @return			number of result bytes, or error code
 */
static int unb64(const char *src, char *dst, size_t dstlen)
{
	char *p;
	unsigned byte1;
	unsigned byte2;
	static char base64[] =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

	if (dstlen < 3)
		return SHORT;

	p = strchr(base64, *src++);

	if (p == NULL)
		return BADCH0;
	byte1 = (p - base64) << 2;	/* first six bits */

	p = strchr(base64, *src++);
	if (p == NULL) {
		return BADCH1;
	}

	byte2 = p - base64;		/* next six:  two plus four */
	*dst++ = byte1 | (byte2 >> 4);
	byte1 = (byte2 & 0xf) << 4;

	p = strchr(base64, *src++);
	if (p == NULL) {
		if (*(src-1) == '=' && *src == '=') {
			if (byte1 != 0)		/* bad padding */
				return BADPAD;
			return 1;
		}
		return BADCH2;
	}

	byte2 = p - base64;		/* next six:  four plus two */
	*dst++ = byte1 | (byte2 >> 2);
	byte1 = (byte2 & 0x3) << 6;

	p = strchr(base64, *src++);
	if (p == NULL) {
		if (*(src-1) == '=') {
			if (byte1 != 0)		/* bad padding */
				return BADPAD;
			return 2;
		}
		return BADCH3;
	}
	byte2 = p - base64;		/* last six */
	*dst++ = byte1 | byte2;

	return 3;
}

/**
 * @brief untext - convert one ASCII character to byte
 * 
 * @param src		known to be full length
 * @param dst		
 * @param dstlen	not large enough is a failure
 * @return 			number of result bytes, or error code
 */
static int untext(const char *src, char *dst, size_t dstlen)
{
	if (dstlen < 1)
		return SHORT;

	*dst = *src;
	return 1;
}

/**
 * @brief badch - produce a nice complaint about an unknown character
 *
 * If the compiler complains that the array bigenough[] has a negative
 * size, that means the TTODATAV_BUF constant has been set too small.
 * 
 * @param src		
 * @param errcode	
 * @param errp		might be NULL
 * @param errlen	
 * @return			literal or errp
 */
static const char *badch(const char *src, int errcode, char *errp, size_t errlen)
{
	static const char pre[] = "unknown character (`";
	static const char suf[] = "') in input";
	char buf[5];
#	define	REQD	(sizeof(pre) - 1 + sizeof(buf) - 1 + sizeof(suf))
	struct sizecheck {
		char bigenough[TTODATAV_BUF - REQD];	/* see above */
	};
	char ch;

	if (errp == NULL || errlen < REQD)
		return "unknown character in input";
	strcpy(errp, pre);
	ch = *(src + BADOFF(errcode));
	if (isprint(ch)) {
		buf[0] = ch;
		buf[1] = '\0';
	} else {
		buf[0] = '\\';
		buf[1] = ((ch & 0700) >> 6) + '0';
		buf[2] = ((ch & 0070) >> 3) + '0';
		buf[3] = ((ch & 0007) >> 0) + '0';
		buf[4] = '\0';
	}
	strcat(errp, buf);
	strcat(errp, suf);
	return (const char *)errp;
}
